content/blog/nielx/2022-12-16_http_network_services_preview_in_r1_beta_4.md

```diff
@@ -0,0 +1,109 @@
+
+The discussion on whether or not exceptions should be used are mostly ideological and fundamental in nature. This preview allows developers to get some hands-on first hand experience of using it in the context of Haiku development, in order to better form their arguments on whether or not the future of the Haiku API is in way of exceptions, or through alternative to `status_t` return values like the [`std::expected`](https://en.cppreference.com/w/cpp/utility/expected) type, which is going to be part of C++23. Or a mixture; there might be good arguments to apply both strategies.
+
+### Thread-safe data access with Exclusive Borrows
+
+One of the biggest challenges in software development is about writing code that handles reading and writing data safely over multiple threads. The main rule is that objects should be owned, read and written by one single thread. As soon as data is shared with more than one thread, there should be mechanisms to synchronize read and write access to the data. The most common methods of synchronization are atomic data structures and semaphores/mutexes. You **need** to employ these synchronization mechanisms, otherwise you will introduce nasty bugs in your code.
+
+In the case of the HTTP Network Services kit, the response body can either be stored as the [`BHttpBody::text`](https://www.haiku-os.org/docs/api/structBPrivate_1_1Network_1_1BHttpBody.html#a2c22caaff871835ac25ecafe3789c4d3) string member, or it can be saved to an object that supports the `BDataIO` interface. When the latter option is used, we must account for the fact that that object will then be shared between the thread that created and scheduled the request, and the thread that processes incoming data and writes this to the target. The `BDataIO` interface itself does not presume or implement any locking mechanism, so that means that there should be an additional system to prevent simultaneous data access. In this case we needed a synchronization mechanism with the following properties:
+
+1. The owner should be able to create an object, and manipulate while they have that in their control.
+2. The owner can then exclusively borrow their object to a borrower, who can than manipulate that object. The borrow is exclusive, meaning that the owner can no longer access or write to the object, until the borrower releases it.
```

Comment:
I'm not native english, but I think the verb "borrow" is used in the other direction (the borrower borrows from the owner). And it could be "the owner lend their object to a borrower"

https://dictionary.cambridge.org/grammar/british-grammar/lend-or-borrow

Reply:
Thanks, will review!

Reply:
Yes, "lend" would be correct here, "borrow" sounds wrong.